README
.. -*- mode: rst -*-

==================================================================
 Helsinki Finite-State Technology (library and application suite)
==================================================================

This package contains a bridging library for multiple FST libraries and toolkits
and set of tools for processing of finite-state automate especially for
linguistic systems. HFST library and tools are licenced under GNU GPL licence
version 3, you may read the full licence in the file named COPYING. The
authors specified in AUTHORS file may be contacted about licencing issues.

There's an online wiki-based information source about HFST in
`Kitwiki HFST pages <https://kitwiki.csc.fi/twiki/bin/view/KitWiki/HfstHome>`_.

Installation
============

Installation method depends on operating system and the version you want to
install. For stable versions there exists packages for some of the better
operating system and package manager combinations. For bleeding edge newest
versions, development and non-supported operating systems and versions you
will have to perform `installation from the sources`_.

Installation packages for Debian and Ubuntu 
-------------------------------------------

Debian packages for HFST are distributed via `Apertium project 
<http://apertium.projectjj.com/apt/nightly/pool/main/h/hfst/>`_. This folder 
contains debian packages for HFST API library and tool packages.
The debian packages are experimental; the requirements of debian or ubuntu
installations are same as main packages. SFST is excluded from the packages
as it has portability issues with hash_maps and hash_sets. The python packages
contain binaries for swig-generated Python bindings for HFST. 

Binaries for Windows
--------------------

Binaries for Windows are distributed via `Apertium project 
<http://apertium.projectjj.com/win32/nightly/>`_. This folder contains
ready-compiled HFST library and command line tools. Performing 
`installation from the sources`_ is also possible on Windows with MinGW 
and Cygwin. Python bindings for Windows are currently available only for 64-bit
Python versions 3.3 and 3.4 on our `download page
<https://hfst.github.io/downloads/index.html>`_.

Installation for MacPorts
-------------------------

The port tree for HFST and related software can be downloaded from `HFST
download directory <https://sourceforge.net/projects/hfst/files/hfst/>`_.
To add this port tree to your macports installation, unpack it to wherever
and insert that directory to your ``sources.conf``, e.g. in the default
installation to ``/opt/local/etc/macports/sources.conf``::

    [...]
    # If an rsync URL points to a .tar file, a signed .rmd160 must exist next to
    # it on the server and will be used to verify its integrity.
    
    file:///Users/flammie/Sources/hfst-macport/
    rsync://rsync.macports.org/release/tarballs/ports.tar [default]

Now you can install hfst and related software using port commands::

  sudo port install hfst

Installation for Gentoo Linux
-----------------------------

HFST software is available in *science* overlay. For portage, use::

  layman -a science

To add the repository. Then::

  emerge -av hfst

to install. Parts of the spell-checking tools are in *Finnish* overlay. If you
use Paludis, just try to ``cave resolve sci-misc/hfst`` to get the current
instructions.

Installation for other systems
------------------------------

For rest of the systems, HFST needs to be installed from the source, the
traditional GNU/linux way. To begin your installation, you need to start by
gathering the dependencies, as adviced in the following chapter.

Dependencies
============

Please note that the dependencies for the library can be set during the compile
time. Settings that determine dependencies relate to features the resulting
library will have; e.g. if you *disable* openfst you cannot use weighted
finite-state automata. If you *enable* foma, you will be able to read foma
format automata as HFST files and use foma's algorithms to process automata. 
The command ``./configure --help`` lists all features that can be controlled
with configure switches and whether they are enabled or disabled by default.

Compilation requirements
------------------------

- To use the OpenFST_ backend (default):

  - source code of OpenFST version 1.2.10 is bundled with HFST and included
    by default when building HFST

  - compiling against OpenFST library and linking may require recent
    GCC version and pthread and m libraries

  - to disable OpenFST support, configure switch ``--without-openfst`` may
    be used (however, this seriously limits the use of HFST)

- To use the SFST_ backend (default):

  - the SFST library version 1.4.6g is bundled with HFST and included
    by default when building HFST

  - SFST requires readline and ncurses

  - The SFST *frontend* a.k.a. the SFST-PL parser a.k.a. ``hfst-sfstpl2fst``
    does **not** require the SFST library to be installed; the library is only
    used for library-stuff like reading SFST format automata.

  - to disable SFST backend, configure switch ``--without-sfst`` must be used

- To use the foma_ backend (default):

  - the foma library version 0.9.18alpha is bundled with HFST and included
    be default when building HFST

  - foma requires editline, termcap (or ncurses) and zlib libraries to compile

  - to disable foma backend, configure switch ``--without-foma`` may be used

  - the *hfst-xfst* frontend does not require foma binary or foma library

- For proper Unicode support in corpus tools:

  - Glib version 2.16 or newer; configure ``--with-unicode-handler=glib``

  - Pkg-config 0.14 or newer

  - Without unicode libraries the corpus tools may be unable to handle
    operations such as *case-folding* properly (default)

  - To compile corpus processing tool ``hfst-proc``, you need to use the
    ``configure`` switch ``--enable-proc`` (or ``--enable-all-tools``)

Note that if you did install dependent libraries, such as libxml or glib
to your home directory instead of using your system's package manager 
(or supported default location):

- If you only have a local version of a library, you can use it with
  appropriate LDFLAGS and CXXFLAGS, eg.
  ``./configure LDFLAGS=-L/path/to/local/lib
  --prefix=/path/to/local/installation`` and
  ``make CXXFLAGS=-I/path/to/local/headers``

If you are building a development version you *loaded from the version control
system*, you must have new brand of GNU development tools installed:

 - autoconf >=2.62

 - automake >=1.11.1

 - libtool >=2.2

 - >=gettext-0.17

 - GNU tool-chain is also needed with distributed packages if the user wishes to
   modify Makefile.am or configure.ac files.

 - Mac OS X users are advised to use MacPorts; Mac OS X 10.6 with
   XCode 2.3.2 at least is not sufficient

 - A package loaded from hfst web site does *not* have these requirements

The source codes *loaded from the version controls system* will also require
parse generator system:

- GNU flex 2.5.35 or compatible and

- GNU bison 1.31 (2.4 suggested) or compatible

- flex 2.5.4-2.5.33 will choke on perfectly valid rules used in hfst

- bison older than 1.31 do not support name-prefix needed for having
  multiple parsers in one library

- A package loaded from hfst web site does *not* have these requirements

- source code loaded from version control system requires them only to
  bootstrap; if you use systems with archaic versions of flex or bison and
  cannot install updates, you might be able to get the needed files from
  somewhere

When running HFST software or using HFST libraries from HFST-enabled software:

- If the executable is dynamically linked (almost always), the operating system
  *must* be able to find hfst libraries

- If you install the libraries in non-standard paths, you need to ensure
  that operating system is aware of this; In linux this may happen by setting
  ``LD_LIBRARY_PATH``, on Mac ``DYLD_LIBRARY_PATH``

- the *hfst-xfst* frontend needs GNU compatible getopt, or basic getopts
  *without* GNU-incompatible getopt installed

- the *hfst-sfstpl2fst* frontend does **not** require SFST libraries or
  binaries, you do **not** need to enable SFST libraries (via the switch 
  --with-sfst) in order to compile SFST-PL scripts to HFST automata.

Installation from the sources
-----------------------------

INSTALL describes the GNU build system in detail, but for most users the usual::

    ./configure
    make
    (as root) make install

should result in a local installation and::

    (as root) make uninstall

in its uninstallation.

If you aren't going to be linking to the library after
building it and don't need to be able to send debugging information, you can
save a considerable amount of space and memory by doing::

    make install-strip

instead of make install. This strips all the symbols from the binaries,
reducing sizes by a factor of 5-10.

If you would rather install in e.g. your home directory
(or aren't the system administrator), you can tell ./configure::

        ./configure --prefix=$HOME

The HFST library may link to numerous FST handling backends with varying
licences. If you are going to redistribute the HFST library you compiled, make
sure there are no clashes in the licences of the linked libraries.

If you are checking out the development versions from SVN you must first create
and install the necessary autotools files from the host system:

  autoreconf -i

It is common practice to keep `generated files out of version control
<http://www.gnu.org/software/automake/manual/automake.html#CVS>`_.

For further installation instruction refer to file ``INSTALL``, which contains
the standard installation instructions for GNU autoconf based software.

If you are compiling HFST from source on Windows with *MinGW*, use the switch 
``--enable-mingw`` when running ``./configure``.
Warning: Using this option with *Cygwin* will cause compilation errors.

Troubleshooting
===============

In this section we list the errors that pop up commonly on `our support channel
<irc://FreeNode/#hfst>`_ or in `our issue tracker
<https://github.com/hfst/hfst/issues>`_,

\::malloc has not been declared
-------------------------------

**During the compilation** errors of form::

  /usr/include/c++/4.3/cstdlib:124: error: '::malloc' has not been declared

or::

  your configure failed to find malloc, check README for further instructions

are mosts often caused by broken library installation. The simplest solution
in Linux-based platforms is ldconfig::

  ldconfig -v

This is actually told and performed by the autotools libtool library
installation, but it's easy to miss. It looks like this::

  Libraries have been installed in:
     /usr/local/lib

  If you ever happen to want to link against installed libraries
  in a given directory, LIBDIR, you must either use libtool, and
  specify the full pathname of the library, or use the '-LLIBDIR'
  flag during linking and do at least one of the following:
     - add LIBDIR to the 'LD_LIBRARY_PATH' environment variable
       during execution
     - add LIBDIR to the 'LD_RUN_PATH' environment variable
       during linking
     - use the '-Wl,-rpath -Wl,LIBDIR' linker flag
     - have your system administrator add LIBDIR to '/etc/ld.so.conf'

  See any operating system documentation about shared libraries for
  more information, such as the ld(1) and ld.so(8) manual pages.

If you installed a library on non-standard path, or installed it to the
default ``/usr/local/lib``, but your variant of Linux doesn't support libraries
there, you may need to set it up and/or ldconfig the directory explicitly::

  export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/usr/local/lib
  ldconfig -v -n /usr/local/lib

If all else fails, try installing the library to wherever your blend of Linux
installs all its libraries, such as /usr/lib.

See also: 
<http://nerdland.net/unstumping-the-internet/malloc-has-not-been-declared/> for
the gory details.

Error while loading shared libraries: libhfst.so.0: cannot open shared object file: No such file or directory
-------------------------------------------------------------------------------------------------------------

**After installing HFST and running programs**, the installed programs should
on most systems be able to find and use the shared libraries that just got
installed alongside the programs that were installed on the same go, but this
is not always the case. Typically on first installation of the HFST library
or after a major version update of HFST library, the system may not know about
it. To fix this, you must run ``ldconfig`` on GNU systems. To ensure proper
linking, use ``ldconfig -v`` to get a print out of known libraries, the
listing should include libhfst.so indicating the current version.

If this is the first time you install a library on your system by hand, it may
happen on some systems that the library configuration does not include that
directory at all. Notably on ubuntu it seems that /usr/local/lib is not one of
library directories, and that is the default target for manually installed
libraries. Either fix this by doing ``./configure --prefix=/usr`` or check
your distributions manuals on how to set up new library directories. Same
applies for libraries installed to e.g. home directory.

See also the previous error description.

chmod: cannot access 'scripts/hfst-foma-wrapper.sh': No such file or directory
------------------------------------------------------------------------------

**During the configuration phase.** This is a case of autoconf being silly; the
commands for creating a script and making it executable from autoconf are ran
in parallel but must be ran once per file and autoconf just doesn't keep track
of the files it creates. The script for doing this is copied to configure.ac
from autoconf manual so it supposably will always work correctly...

fatal error: htwolcpre1.h: No such file or directory
----------------------------------------------------

**During the compilation phase**, an error message including things like::

 scanner1.ll:22:27: fatal error: htwolcpre1.h: No such file or directory
 compilation terminated.
 Makefile:1029: recipe for target `scanner1.o' failed

indicates that the header files generated by flex/yacc have the extension `.hh` 
instead of `.h`. This error has been encountered at least on some versions of 
Cygwin. Currently, the best solution is to manually rename each
flex/yacc-generated header file of the form `foo.hh` as `foo.h`. You probably
have to do this iteratively after each error message of this type.

....libs/lt-hfst-strings2fst: Incorrect utf-8 coding
----------------------------------------------------

**During the make check phase**. This indicates that a test that tests for
expected failure fails expectedly, which is also indicated by a green word
*PASS*  or *XFAIL* on the next line. This is expected behaviour and not an
indication of a bug. If there is a bug effecting that or other tests in the
test suite, it will be indicated by a line starting with word *FAIL* or *XPASS*
in red colour.

Same applies for other messages printed during make check phase. The cases where
something actually fails will be clearly stated in the end of the test suite by
a message such as::

  ======================================
  2 of 36 tests failed
  Please report to hfst-bugs@helsinki.fi
  ======================================

These errors can be reported either to the stated mail address or the `HFST
bug tracker in sf.net infrastructure
<https://sourceforge.net/tracker/?atid=1061990&group_id=224521&func=browse>`_.

cat: hvVqf:o:l:u:: No such file or directory
--------------------------------------------
**During use of bash-based scripts**, an error message including things like::


  cat: -l: No such file or directory
  cat: version,quiet,format:,output:,latin1::,utf8::: No such file or directory
  cat: -n: No such file or directory
  cat: hfst-lexc: No such file or directory
  cat: --: No such file or directory

indicate that script is trying to use Mac OS X's getopt as if it was GNU getopt.
However default getopt in Mac OS X does not work at all like GNU getopt.
Easiest solution is to install working getopt, e.g. by using MacPorts::

  sudo port install getopt

The newer versions of bash scripts detect Mac OS X's getopt and fallback to
using getopts. Note that getopts does not support long options and filenames
must be last parameters on commandline with it, so its use is strongly
discouraged.

libc++-abi.dylib: terminate called throwing an exception
--------------------------------------------------------

**During program execution** *(Mac OS X only)*, errors of form::

  terminate called throwing an exception
  Abort trap: 6

Can be caused by, just about any exceptional situation that does not have
specific handler. On Linux it will read::

  terminate called after throwing an instance of 'ImplementationTypeNotAvailableException'

And then you'll know that this specific exception is about backend that was disabled during `configure` phase. Or it might read::

 terminate called after throwing an instance of 'UndefinedSymbolPairsFound'

And you'd know it's something with the alphabet. But OS X won't tell us this. So
it is an unexpected error situation. Usual suspects are still: missing library
in configure, empty file, reading error, writing error...


Further information
===================

The `Hfst wiki site
<https://kitwiki.csc.fi/twiki/bin/view/KitWiki/HfstHome>`_ contains further
details of the HFST system.

Reporting bugs
==============

Bugs can be reported via email to `HFST team bug mail address
<hfst-bugs@helsinki.fi>`_, or preferably to `HFST's bug tracking system
at Github
<https://github.com/hfst/hfst/issues>`_
When reporting, please include at least following:

* version of software used, if command-line tool (hfst-toolname --version)

* version of hfst-library, if possible

* steps to reproduce, attach or all related files if possible

* information about platform used (e.g. uname -a)

.. _libxml2: http://www.xmlsoft.org/
.. _libreadline: http://www.gnu.org/software/readline/

.. vim: set ft=rst:


NEWS
=================================
 NEWS for hfst-tools and library
=================================

This file contains all noteworthy changes in HFST development between releases.
For full listing of changes see ChangeLog.

Noteworthy changes in 3.12.2
----------------------------

* Changes to configure:

  * Disable lexc and foma wrappers as well as hfst-train-tagger tool unless explicitly requested

  * Enable hfst-calculate and hfst-xfst by default

  * Add experimental options --with-openfst-log=lean and --with-sfst=lean which support a limited nuber of operations for these types (reading, writing, converting between types and deleting)

  * Require at least automake version 1.12 unless compiling from pre-yacc-generated sources

* Implement hfst-twolc as a single program instead of a script. Get rid of hfst-twolc-system and hfst-twolc-loc tools.

* Improve pmatch compilation and error handling mechanisms

* Improve hfst-tokenize tool

* Add transliterate output mode (--transliterate) to hfst-proc

* Changes to python interface:

  * Support twolc and sfst compilation

  * Improve HfstBasicTransducer iteration mechanism

  * Add experimental support for python version 2

  * Support apply up and apply down commands in function start_xfst


Noteworthy changes in 3.12.1
----------------------------

* Fix flag elimination bug (reported in issue #342)

* Do not allow unescaped dots in regular expressions

* Improvements to pmatch and tokenization tools:

  * Search for included files under scriptdir, not working dir

  * Add experimental two-vector model for word sense

  * Handle Apertium-style superblanks in --giella-cg format

* Rename the PyPI package to 'hfst' (currently available at testpypi.python.org)

* Update foma back-end


Noteworthy changes in 3.12.0
----------------------------

* fixes to memory leaks and efficiency

* fixes to numerous warnings

* changes in Python bindings:

  * rename hfst.rules into hfst.sfst_rules

  * get rid of hfst.types and offer implementation types in class ImplementationType

  * add Xerox-type rules in module hfst.xerox_rules

  * improve documentation

  * tentatively add partial support for pypi installation

* improvements to pmatch tools and hfst-proc


Noteworthy changes in 3.11.0
----------------------------

* Add docstrings to Python API

* Changes and improvements to pmatch tools, hfst-tokenize, hfst-optimized-lookup, hfst-lookup and hfst-xfst

* Fix bugs in flag elimination


Noteworthy changes in 3.10.0
----------------------------

* Swap directions of 'apply up' and 'apply down' in hfst-xfst, so that these commands work in the same way as in foma and xfst.

* Add a new tool called hfst-flookup. It does lookup from right to left, in the same way as foma's flookup and xerox's lookup. The tools hfst-lookup and hfst-optimized-lookup stay as they are.

* Improvements to pmatch and optimized lookup

* Make hfst-fst2fst print a more informative error message if a gzipped native foma transducer is given as input.

* Changes and improvements in Python interface:

  * Transducer functions returning a reference to the transducer now return void.

  * Tentatively add some pmatch functions.

  * Add some two-level rule functions.

  * Support lookup for transducers not in optimized-lookup format.

  * Support Python 2 unicode strings in lookup.


Noteworthy changes in 3.9.2
---------------------------

* Improvements to tokenization and pmatch tools


Noteworthy changes in 3.9.1
---------------------------

* License for the library code (under libhfst and python directories) is now LGPL 3

* Make python bindings compatible also with python2

* Add time cutoff option to hfst-pmatch and hfst-proc2

* Rename hfst-proc2 to hfst-tokenize, with hfst-proc2 and hfst-tokenise as aliases

* Use standard c++0x when compiling unless a higher standard is required with -std

* Check automake version during configuration to determine whether header files generated by yacc/bison should use extension .h (automake < 1.12) or .hh (automake >= 1.12)

* Omit infinitely ambiguous check in hfst-lookup when time cutoff is defined

* Improvements in Xerox markup rules

* A bugfix in compile-replace in hfst-xfst

* Improvements in pmatch tools, hfst-optimized-lookup, hfst-reweight, hfst-pair-test and hfst-substitute

* A bugfix in priority union


Noteworthy changes in 3.9.0
---------------------------

* Implement a new Python API

* Add option --time-cutoff to hfst-lookup

* Remove option --minimizeFlags from hfst-lexc and minimize flags by default, add option --alignStrings

* Add support for controlling input and output streams in xre, xfst and lexc compilers

* Update foma back-end

* Improve handling of negative weights, check for epsilon cycles with negative weights and issue a warning, if needed

* Improvements and bug fixes to hfst-optimized-lookup, hfst-proc2, hfst-proc and hfst-pmatch

* Reduce memory leaks when using hfst-ol format

* Support flag diacritics in replace rules

* Rewrite markup rules

* Make command line tools perform automatic conversion when processing transducers whose types differ


Noteworthy changes in 3.8.3
---------------------------

* Bugfixes and improvements to many tools, including hfst-xfst, hfst-proc(2),
  hfst-lexc, hfst-pmatch(2fst), hfst-reweight, hfst-pair-test, hfst-compose-intersect

* Better support for command line tools on windows


Noteworthy changes in 3.8.2
---------------------------

* Runtime speed improvements in pmatch

* Speedup in conversion to optimized lookup format

* Merge and compile-replace operations supported in hfst-xfst

* Added option --print-symbol-pair-statistis(=N) to hfst-summarize


Noteworthy changes in 3.8.1
---------------------------

* (the release of version 3.8.0 was delayed, its changes are also listed here)

* changes to flag diacritic handling

  * now hfst-lexc has options --withFlags, --minimizeFlags and --renameFlags to control
    hyperminimization


python/pypi_setup.py
#!/usr/bin/python3

"""
Setup for creating PIP packages for HFST Python bindings.

Before running setup, recursively copy directories 'libhfst/src'
and 'back-ends' from HFST c++ source code under the directory where
setup is run. Make sure that the following c++ and header files 
from 'libhfst/src/parsers' have been generated from flex/yacc
files before copying:

  lexc-parser.cc pmatch_parse.cc xfst-parser.cc xre_parse.cc
  lexc-parser.hh pmatch_parse.hh xfst-parser.hh xre_parse.hh

Compiling the extensions requires python, swig and a c++ compiler, 
all located on a directory listed on system PATH. On linux and mac 
osx, readline and getline must be available and the c++ compiler
must support flag 'std=c++0x'.

The setup script has been tested on linux with gcc 4.6.3 and
swig 2.0.4 (with python 3.2mu and 3.4m) and on windows with
msvc 10.0 and swig 3.0.5 (with python 3.3. and 3.4).

"""

import os
# from distutils.core import setup, Extension
from setuptools import setup, Extension
from sys import platform

# HFST C++ headers needed by swig when creating the python/c++ interface
swig_include_dir = "libhfst/src/"

ext_swig_opts = ["-c++", "-I" + swig_include_dir, "-Wall"]
if platform == "win32":
    ext_swig_opts = ["-c++", "-I" + swig_include_dir, "-Wall", 
                     "-IC:\\Program Files (x86)\\Microsoft SDKs\\Windows\\v7.0A\\Include"]
# todo: see if SDK directory is needed

# readline is needed for hfst.start_xfst(), on windows the shell where HFST
# python bindings are run from has its own readline which will do
ext_extra_link_args = []
if platform == "linux" or platform == "linux2" or platform == "darwin":
    ext_extra_link_args = ['-lreadline']

# HFST headers needed when compiling the actual c++ extension
ext_include_dirs = [".", "libhfst/src/", "back-ends/foma", "back-ends",
                    "parsers", "libhfst/src/parsers"]
if platform == "win32":
    ext_include_dirs.append("back-ends/openfstwin/src/include")
else:
    ext_include_dirs.append("back-ends/openfst/src/include")

# this replaces ./configure
ext_define_macros = [ ('HAVE_FOMA', None), ('HAVE_OPENFST', None),
                      ('HAVE_OPENFST_LOG', None) ]
if platform == "linux" or platform == "linux2" or platform == "darwin":
    ext_define_macros.append(('HAVE_READLINE', None))
    ext_define_macros.append(('HAVE_GETLINE', None))
if platform == "win32":
    for macro in ["HFSTEXPORT", "OPENFSTEXPORT", "_MSC_VER", "WINDOWS", "WIN32", "_CRT_SECURE_NO_WARNINGS"]:
        ext_define_macros.append((macro, None))

# use c++0x standard, if possible
ext_extra_compile_args = ['']
if platform == "linux" or platform == "linux2" or platform == "darwin":
    ext_extra_compile_args = ["-std=c++0x", "-Wno-sign-compare", "-Wno-strict-prototypes"]
# define error handling mechanism on windows
if platform == "win32":
    ext_extra_compile_args = ["/EHsc"]

ext_library_dirs = []
ext_libraries = []

# on windows, c++ source files have 'cpp' extension
cpp = ".cc"
if platform == "win32":
    cpp = ".cpp"

# on windows, openfst back-end is in directory 'openfstwin'
openfstdir = "openfst"
if platform == "win32":
    openfstdir = "openfstwin"

# all c++ extension source files
libhfst_source_files = ["libhfst/src/parsers/XfstCompiler" + cpp,
                        "libhfst/src/HfstApply" + cpp,
                        "libhfst/src/HfstInputStream" + cpp,
                        "libhfst/src/HfstTransducer" + cpp,
                        "libhfst/src/HfstOutputStream" + cpp,
                        "libhfst/src/HfstRules" + cpp,
                        "libhfst/src/HfstXeroxRules" + cpp,
                        "libhfst/src/HfstDataTypes" + cpp,
                        "libhfst/src/HfstSymbolDefs" + cpp,
                        "libhfst/src/HfstTokenizer" + cpp,
                        "libhfst/src/HfstFlagDiacritics" + cpp,
                        "libhfst/src/HfstExceptionDefs" + cpp,
                        "libhfst/src/HarmonizeUnknownAndIdentitySymbols" + cpp,
                        "libhfst/src/HfstLookupFlagDiacritics" + cpp,
                        "libhfst/src/HfstEpsilonHandler" + cpp,
                        "libhfst/src/HfstStrings2FstTokenizer" + cpp,
                        "libhfst/src/HfstPrintDot" + cpp,
                        "libhfst/src/HfstPrintPCKimmo" + cpp,
                        "libhfst/src/hfst-string-conversions" + cpp,
                        "libhfst/src/string-utils" + cpp,
                        "libhfst/src/implementations/HfstBasicTransducer" + cpp,
                        "libhfst/src/implementations/HfstBasicTransition" + cpp,
                        "libhfst/src/implementations/ConvertTransducerFormat" + cpp,
                        "libhfst/src/implementations/HfstTropicalTransducerTransitionData" + cpp,
                        "libhfst/src/implementations/ConvertTropicalWeightTransducer" + cpp,
                        "libhfst/src/implementations/ConvertLogWeightTransducer" + cpp,
                        "libhfst/src/implementations/ConvertFomaTransducer" + cpp,
                        "libhfst/src/implementations/ConvertOlTransducer" + cpp,
                        "libhfst/src/implementations/TropicalWeightTransducer" + cpp,
                        "libhfst/src/implementations/LogWeightTransducer" + cpp,
                        "libhfst/src/implementations/FomaTransducer" + cpp,
                        "libhfst/src/implementations/HfstOlTransducer" + cpp,
                        "libhfst/src/implementations/compose_intersect/ComposeIntersectRulePair" + cpp,
                        "libhfst/src/implementations/compose_intersect/ComposeIntersectLexicon" + cpp,
                        "libhfst/src/implementations/compose_intersect/ComposeIntersectRule" + cpp,
                        "libhfst/src/implementations/compose_intersect/ComposeIntersectFst" + cpp,
                        "libhfst/src/implementations/compose_intersect/ComposeIntersectUtilities" + cpp,
                        "libhfst/src/implementations/optimized-lookup/transducer" + cpp,
                        "libhfst/src/implementations/optimized-lookup/convert" + cpp,
                        "libhfst/src/implementations/optimized-lookup/ospell" + cpp,
                        "libhfst/src/implementations/optimized-lookup/pmatch" + cpp,
                        "libhfst/src/implementations/optimized-lookup/find_epsilon_loops" + cpp,
                        "libhfst/src/parsers/xre_lex" + cpp,
                        "libhfst/src/parsers/xre_parse" + cpp,
                        "libhfst/src/parsers/pmatch_parse" + cpp,
                        "libhfst/src/parsers/pmatch_lex" + cpp,
                        "libhfst/src/parsers/lexc-parser" + cpp,
                        "libhfst/src/parsers/lexc-lexer" + cpp,
                        "libhfst/src/parsers/xfst-parser" + cpp,
                        "libhfst/src/parsers/xfst-lexer" + cpp,
                        "libhfst/src/parsers/LexcCompiler" + cpp,
                        "libhfst/src/parsers/PmatchCompiler" + cpp,
                        "libhfst/src/parsers/XreCompiler" + cpp,
                        "libhfst/src/parsers/lexc-utils" + cpp,
                        "libhfst/src/parsers/pmatch_utils" + cpp,
                        "libhfst/src/parsers/xre_utils" + cpp,
                        "libhfst/src/parsers/xfst-utils" + cpp,
                        "libhfst/src/parsers/xfst_help_message" + cpp ]

foma_source_files = [ "back-ends/foma/int_stack.c",
                      "back-ends/foma/define.c",
                      "back-ends/foma/determinize.c",
                      "back-ends/foma/apply.c",
                      "back-ends/foma/rewrite.c",
                      "back-ends/foma/topsort.c",
                      "back-ends/foma/flags.c",
                      "back-ends/foma/minimize.c",
                      "back-ends/foma/reverse.c",
                      "back-ends/foma/extract.c",
                      "back-ends/foma/sigma.c",
                      "back-ends/foma/structures.c",
                      "back-ends/foma/constructions.c",
                      "back-ends/foma/coaccessible.c",
                      "back-ends/foma/io.c",
                      "back-ends/foma/utf8.c",
                      "back-ends/foma/spelling.c",
                      "back-ends/foma/dynarray.c",
                      "back-ends/foma/mem.c",
                      "back-ends/foma/stringhash.c",
                      "back-ends/foma/trie.c",
                      "back-ends/foma/lex.yy.c",
                      "back-ends/foma/regex.c" ]

openfst_source_files =  [ "back-ends/" + openfstdir + "/src/lib/compat" + cpp,
                          "back-ends/" + openfstdir + "/src/lib/flags" + cpp,
                          "back-ends/" + openfstdir + "/src/lib/fst" + cpp,
                          "back-ends/" + openfstdir + "/src/lib/properties" + cpp,
                          "back-ends/" + openfstdir + "/src/lib/symbol-table" + cpp,
                          "back-ends/" + openfstdir + "/src/lib/symbol-table-ops" + cpp,
                          "back-ends/" + openfstdir + "/src/lib/util" + cpp ]

# todo: see what files are actually needed...
# if platform == "linux" or platform == "linux2" or platform == "darwin":
#    foma_source_files.append("back-ends/foma/lexcread.c")
#    foma_source_files.append("back-ends/foma/lex.lexc.c")

libhfst_source_files = libhfst_source_files + openfst_source_files

if platform == "linux" or platform == "linux2" or platform == "win32":
    libhfst_source_files = libhfst_source_files + foma_source_files

# clang doesn't accept "-std=c++0x" flag when compiling C,
# so foma back-end must be compiled separately
foma_object_files = []
if platform == "darwin":
    import subprocess
    for file in foma_source_files:
        subprocess.call(["gcc", "-fPIC", "-std=c99", "-c", file])
        foma_object_files.append(file.replace('back-ends/foma/','').replace('.c','.o'))

# The HFST c++ extension
libhfst_module = Extension('_libhfst',
                           language = "c++",
                           sources = ["libhfst.i"] + libhfst_source_files,
                           swig_opts = ext_swig_opts,
                           include_dirs = ext_include_dirs,
                           library_dirs = ext_library_dirs,
                           libraries = ext_libraries,
                           define_macros = ext_define_macros,
                           extra_link_args = ext_extra_link_args,
                           extra_compile_args = ext_extra_compile_args,
                           extra_objects = foma_object_files
                           )

setup(name = 'hfstpy',
      version = '3.12.0_beta',
      author = 'HFST team',
      author_email = 'hfst-bugs@helsinki.fi',
      url = 'http://hfst.github.io/',
      description = 'SWIG-bound hfst interface',
      license = 'GNU GPL3',
      ext_modules = [libhfst_module],
      py_modules = ["libhfst"],
      packages = ["hfst", "hfst.exceptions", "hfst.sfst_rules", "hfst.xerox_rules"],
      data_files = []
      )
